pubsub/streaming-analytics/PubSubToGCS.py

```diff
@@ -14,119 +14,126 @@
 
 # [START pubsub_to_gcs]
 import argparse
-import datetime
-import json
+from datetime import datetime
 import logging
+import random
 
-import apache_beam as beam
+from apache_beam import DoFn, GroupByKey, io, ParDo, Pipeline, PTransform, WindowInto
 from apache_beam.options.pipeline_options import PipelineOptions
-import apache_beam.transforms.window as window
+from apache_beam.transforms.window import FixedWindows
 
 
-class GroupWindowsIntoBatches(beam.PTransform):
-    """A composite transform that groups Pub/Sub messages based on publish
-    time and outputs a list of dictionaries, where each contains one message
-    and its publish timestamp.
+class GroupMessagesByFixedWindows(PTransform):
+    """A composite transform that groups Pub/Sub messages based on publish time
+    and outputs a list of tuples, each containing a message and its publish time.
     """
 
-    def __init__(self, window_size):
-        # Convert minutes into seconds.
+    def __init__(self, window_size, num_shards):
+        # Set window size to 60 seconds.
         self.window_size = int(window_size * 60)
+        self.num_shards = num_shards
 
     def expand(self, pcoll):
         return (
             pcoll
-            # Assigns window info to each Pub/Sub message based on its
-            # publish timestamp.
-            | "Window into Fixed Intervals"
-            >> beam.WindowInto(window.FixedWindows(self.window_size))
-            | "Add timestamps to messages" >> beam.ParDo(AddTimestamps())
-            # Use a dummy key to group the elements in the same window.
-            # Note that all the elements in one window must fit into memory
-            # for this. If the windowed elements do not fit into memory,
-            # please consider using `beam.util.BatchElements`.
-            # https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.util.html#apache_beam.transforms.util.BatchElements
-            | "Add Dummy Key" >> beam.Map(lambda elem: (None, elem))
-            | "Groupby" >> beam.GroupByKey()
-            | "Abandon Dummy Key" >> beam.MapTuple(lambda _, val: val)
+            # Bind window info to each element using element timestamp (or publish time).
+            | "Window into fixed intervals"
+            >> WindowInto(FixedWindows(self.window_size))
+            | "Add key to windowed elements" >> ParDo(AddKey(self.num_shards))
```

Comment:
I would actually add the timestamp in one transform, and add the random key in another transform. It's usually clearer to have each transform do only one simple function. Dataflow fuses all the element-wise transforms together into a single function anyways, so there are no performance concerns from this.

```
  | 'Add timestamps' >> beam.ParDo(AddTimestamps())
  | 'Random shard keys' >> beam.WithKeys(lambda _: random.randint(0, num_shards - 1))
```

It would also be a good idea to explain at some point _why_ we are doing sharding, what shards are, and how to choose a good number of shards. Maybe the README would be a good place for this?

Reply:
Moved this out of this DoFn and mentioned sharding in README.